torch/_dynamo/variables/higher_order_ops.py

```diff
@@ -500,22 +500,64 @@ def speculate_branch(branch):
         false_nn_modules = tx.copy_graphstate().output.nn_modules
 
         def dedup_and_sort_lifted_freevars(true_lifted_freevars, false_lifted_freevars):
-            shared_freevars = true_lifted_freevars.keys() & false_lifted_freevars.keys()
-            unique_true_freevars = true_lifted_freevars.keys() - shared_freevars
-            unique_false_freevars = false_lifted_freevars.keys() - shared_freevars
+            # The nn module attributes are guaranteed to be registered into the top-level graph module during
+            # higher order op speculation. Therefore, get_attr nodes in two branches with the same
+            # target refer to the same attribute and we can safely deduplicate them with their target.
```

Comment:
Do you have an example of a before graph (with the true/false branches) and an after graph? 
Also, do the proxies belong to the SubgraphTracers of the true_fn and false_fn, or do they belong to the outer SubgraphTracer?

Reply:
The proxies belong to the root_tracer. Related code is [here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/output_graph.py#L709). 

The attribute was registered twice: once in true_branch and once in false_branch. builder uses side effect object to avoid registering the same attribute twice see [here](https://github.com/pytorch/pytorch/blob/main/torch/_dynamo/variables/builder.py#L231-L242)). However, we restore the side-effects after each speculation. This makes builder to call register_attr_or_module again and it creates a new proxy there.


I've also added an example before and after the change in the PR description.